File: src/nodejs/src/mappers/PatientMapper.ts
import { Patient } from "../entities/Patient";
import { PatientDto } from "../dtos/PatientDto";
import { OrderDto } from "../dtos/OrderDto";

/**
 * Mapper for Patient related objects.
 */

// Extend the Patient interface to add the toDto method
declare module "../entities/Patient" {
  interface Patient {
    toDto(): PatientDto;
  }
}

/**
 * Maps a Patient entity to a PatientDto.
 * @param patient The patient entity.
 * @returns The patient DTO.
 */
Patient.prototype.toDto = function (): PatientDto {
  const ordersDto = this.orders?.map((o) => new OrderDto(o.id, o.message, o.patientId, o.createdAt, o.updatedAt)) || [];

  return new PatientDto(this.id, this.name, this.createdAt, ordersDto, this.updatedAt);
};

/**
 * Maps a collection of Patient entities to a collection of PatientDtos.
 * @param patients The collection of patient entities.
 * @returns The collection of patient DTOs.
 */
export function toDtos(patients: Patient[]): PatientDto[] {
  return patients.map((p) => p.toDto());
}

File: src/csharp/api/Mappers/PatientMapper.cs
using Example.Api.Dtos;
using Example.Api.Models;

namespace Example.Api.Mappers;

/// <summary>
/// Mapper for Patient related objects.
/// </summary>
public static class PatientMapper
{
    /// <summary>
    /// Maps a Patient entity to a PatientDto.
    /// </summary>
    /// <param name="patient">The patient entity.</param>
    /// <returns>The patient DTO.</returns>
    public static PatientDto ToDto(this Patient patient)
    {
        return new PatientDto
        {
            Id = patient.Id,
            Name = patient.Name,
            Age = patient.Age,
            Gender = patient.Gender,
            Email = patient.Email,
            PhoneNumber = patient.PhoneNumber,
            DateOfBirth = patient.DateOfBirth,
            Address = patient.Address,
            FirstVisitDate = patient.FirstVisitDate,
            Status = patient.Status,
            Remarks = patient.Remarks,
            CreatedAt = patient.CreatedAt,
            UpdatedAt = patient.UpdatedAt,
            Orders = patient.Orders?.ToDtos().ToList() ?? Enumerable.Empty<PatientOrderDto>().ToList(),
        };
    }

    /// <summary>
    /// Maps a collection of Patient entities to a collection of PatientDtos.
    /// </summary>
    /// <param name="patients">The collection of patient entities.</param>
    /// <returns>The collection of patient DTOs.</returns>
    public static IEnumerable<PatientDto> ToDtos(this IEnumerable<Patient> patients)
    {
        return patients.Select(p => p.ToDto());
    }
}
